src/components/TheViz.vue

```diff
@@ -0,0 +1,257 @@
+  project,
+  getMidpoint,
+} from "@/util/math";
+
+// settings shared across functions
+const size = 4; // size of dots and links
+const gap = 50; // distance between dots
+// https://pinetools.com/blend-colors
+const baseColor: Color = [25, 143, 154]; // color of elements at rest (10% white, 90% theme-dark)
+const pulseColor: Color = [127, 193, 199]; // color of element when pulsing (50% white, 50% theme-dark)
+// (other one-off settings in place below)
+
+// 3d axis rotations
+let rx = 0;
+let ry = 0;
+let rxTarget = 0;
+let ryTarget = 0;
+
+// efficient rgb tuple format
+type Color = [number, number, number];
+
+interface Dot {
+  // position in 3d space
+  point: Point3d;
+  // 3d position projected into 2d for canvas rendering
+  projected?: Point2d;
+  // target and actual color
+  color: Color;
+  colorTarget: Color;
+}
+
+interface Link {
+  // dots to link together
+  from: Dot;
+  to: Dot;
+  // target and actual color (in efficient rgb tuple format)
+  color: Color;
+  colorTarget: Color;
+}
+
+// globals
+let canvas = null as HTMLCanvasElement | null;
+let ctx = null as CanvasRenderingContext2D | null;
+let width = 0;
+let height = 0;
+let dots: Array<Dot> = [];
+let links: Array<Link> = [];
+
+// resize canvas
+const resize = () => {
+  if (!canvas || !ctx) return;
+  const scale = window.devicePixelRatio;
+  width = canvas.clientWidth;
+  height = canvas.clientHeight;
+  canvas.width = width * scale;
+  canvas.height = height * scale;
+  ctx.scale(scale, scale);
+};
+
+// generate field of dots and links
+const generate = debounce(() => {
+  // generate field of dots
+  dots = [];
+  // start from grid so dots relatively uniformly distributed
+  for (let x = -gap; x <= width + gap; x += gap) {
+    for (let y = -gap; y <= height + gap; y += gap) {
+      // eliminate some to make more "organic" and less "grid"
+      if (Math.random() > 0.5) {
+        const angle = Math.random() * 360;
+        dots.push({
+          point: {
+            // nudge off grid to make more "organic"
+            x: x + sin(angle) * (gap / 4),
+            y: y + cos(angle) * (gap / 4),
+            // random range to create nice thin-ish 3d layer
+            z: -gap + Math.random() * 2 * gap,
+          },
+          color: [...baseColor],
+          colorTarget: [...baseColor],
+        });
+      }
+    }
+  }
+
+  // hard limit dots for performance
+  while (dots.length > 200)
+    dots.splice(Math.floor(dots.length * Math.random()), 1);
+
+  // sort dots by z
+  dots.sort((a, b) => a.point.z - b.point.z);
+
+  // go through each pair of dots
+  links = [];
+  for (let a = 0; a < dots.length; a++) {
+    for (let b = 0; b < dots.length; b++) {
+      // upper triangular matrix to only count each pair once
+      if (a > b) {
+        const from = dots[a];
+        const to = dots[b];
+        // only link if dots close enough together
+        if (
+          dist(from.point.x, from.point.y, to.point.x, to.point.y) <
+          gap * 1.5
```

Comment:
Since, you're interested in optimization, you might consider adding a `distSq()` function that doesn't call `Math.sqrt()`, which is typically slow. You'd replace the above with something like `distSq(...) < Math.pow(gap * 1.5, 2)`.

I don't know how much of a difference it makes nowadays, but it was a popular trick back in the day.

Reply:
Cool trick, I've never heard of that one. I'll keep it in mind for the future if we find it running unacceptably slow on an old device, but I imagine v8 and other js engines are probably so optimized today that it's not needed? Also sub 1ms is good with me, and dist is only a small part of everything being done. Note that it's only being called in `generate()` (which only runs once on page load or resize, debounced) and `pulse()` (which runs every once 5 seconds). So given that I'd like to keep it the recognizable distance formula.

\* "once" one frame, once per dot.